src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Workspace/Mef/MefWorkspaceServices.cs

```diff
@@ -51,19 +57,53 @@ public override HostServices HostServices
 
         public override Workspace Workspace => _workspace;
 
+        public override void Dispose()
+        {
+            ImmutableArray<IDisposable> disposableServices;
+            lock (_gate)
+            {
+                disposableServices = _ownedDisposableServices.ToImmutableArray();
+                _ownedDisposableServices.Clear();
+            }
+
+            foreach (var service in disposableServices)
+            {
+                try
+                {
+                    service.Dispose();
+                }
+                catch (Exception ex) when (FatalError.ReportWithoutCrashUnlessCanceled(ex))
+                {
+                    // Ignored
+                }
+            }
+
+            base.Dispose();
+        }
+
         public override TWorkspaceService GetService<TWorkspaceService>()
         {
             if (TryGetService(typeof(TWorkspaceService), out var service))
             {
-                return (TWorkspaceService)service.Value;
+                var checkAddDisposable = service.usesFactory && !service.lazyService.IsValueCreated;
+                var result = (TWorkspaceService)service.lazyService.Value;
+                if (checkAddDisposable && result is IDisposable disposable)
```

Comment:
Only if the return value is also disposable. Previously disposing of these cases was not supported so I would not expect it to have any impact.

Reply:
I guess I have little faith somebody wouldn't have returned a disposable object without realizing it. We audited at least all the workspace factory exports in Roslyn and a partner repos to make sure nobody made that mistake?

Reply:
We haven't audited, but the main risk mitigation here is we do not dispose workspaces in production scenarios.

Reply:
Prior to this change, there are no workspace service factories in dotnet/roslyn that return a service instance that implements `IDisposable` (this counts both singletons and per-instance returns).

Reply:
> Prior to this change, there are no workspace service factories in dotnet/roslyn that return a service instance that implements IDisposable (this counts both singletons and per-instance returns).

Right.  But my point is: can we solve this *without* the heavyweight complexity of having to thread through the IDispsoable pattern.  We're only do that to trigger a single CTS.  If tehre was a simple service one could just import if they needed to know when teh Workspace is going away, then it feels like we woudln't need this.

Reply:
> We're only do that to trigger a single CTS.

This is a bit of a red herring. See efcb2b5a8554cc0c3aab61a1409331e4a98d43f7 and ec884a0289619c1280719b1d854a7705df00935e for other examples.

> If tehre was a simple service one could just import if they needed to know when teh Workspace is going away, then it feels like we woudln't need this.

The MEF catalog already implements support for `IDisposable`. If we consider that `MefWorkspaceServices` is really a scope-limited composition, then this change is conceptually as simple as it gets. MEF-based exports are tied to an owner (whether that's the `ExportProvider` or otherwise) and are disposed if necessary when the owner is disposed.